Docs: Improve documentation for notifications (#59504)
grafana/grafana: docs/sources/alerting/fundamentals/notifications/index.md

  - contact point
  - notification channel
  - create
title: Notifications
weight: 410
---

# Notifications

> Notification are sent when an alert is firing or has been resolved, and configuring how and where a notification should be sent is done via notification policies. With notification policies it is possible to configure how often a notification should be sent; and whether alerts should all be sent in the same notification, sent in grouped notifications based on a set of labels, or as seperate notifications.

Author: Discussed this with Brenda. I want to make sure that there is some introduction to notification policies, but the information in that introduction isn't lost if someone chooses to skip the introduction and go straight to reading `Notification policies`.

## Notification policies

Notification policies control when and where notifications are sent. A notification policy can choose to send all alerts together in the same notification, send alerts in grouped notifications based on a set of labels, or send alerts as separate notifications. Each notification policy can have its own timings configuration that control how often notifications should be sent; and can reference one or more mute timings to inhibit notifications at certain times of the day and on certain days of the week.

> Notification policies are organized in a tree structure where at the root of the tree there is a notification policy called the root policy. There can be only one root policy, and the root policy cannot be deleted. The root policy also happens to be where all firing and resolved alerts are sent.

Author: Hmm. What I'm attempting to explain here is that the root policy is always the starting place for policy matching, and I'm trying to figure out how to word it.

> Notification policies are organized in a tree structure where at the root of the tree there is a notification policy called the root policy. There can be only one root policy, and the root policy cannot be deleted. The root policy also happens to be where all firing and resolved alerts are sent.

Reviewer: > The root policy is always the starting place for policy matching.

Maybe you just did? I like that wording! 😆

"The root policy is always the starting place for determining where to send notifications." A little closer to the original?

Specific routing policies are children of the root policy, and can be used to match either all alerts, or a subset of alerts based on a set of labels. An example of a specific routing policy could be sending high priority alerts to Pagerduty and sending low priority alerts as emails. Another example of a specific routing policy could be sending infrastructure alerts to the Ops team, and all other alerts to another team.

When an alert is sent to the root policy it looks at each specific routing policy and sends the alert to the first specific routing policy that matches the alert. If there are no specific routing policies, or none which match the alert, then the root policy is the matching policy.

> 

Author: I think I am going to put this in the more advanced documentation for notification policies, this page is just an introduction to them.

More information on how to configure notification policies can be found [here]({{< relref "../../manage-notifications/create-notification-policy/" >}}).

## Contact points

> Contact points contain the configuration for sending notifications. A contact point is a list of integrations each of which sends a notification to a particular email address, service or URL. Contact points can have multiple integrations of the same kind, or a combination of integrations of different kinds. For example, a contact point could contain a Pagerduty integration; an email and Slack integration; or a Pagerduty integration, a Slack integration, and two email integrations. A contact point can also be configured with no integrations in which case no notifications are sent.

Author: Should it be `matching alerts`?

> Contact points contain the configuration for sending notifications. A contact point is a list of integrations each of which sends a notification to a particular email address, service or URL. Contact points can have multiple integrations of the same kind, or a combination of integrations of different kinds. For example, a contact point could contain a Pagerduty integration; an email and Slack integration; or a Pagerduty integration, a Slack integration, and two email integrations. A contact point can also be configured with no integrations in which case no notifications are sent.

Reviewer: I originally wrote that, but went back on it because technically an alert is not sent, a notification is sent. But it probably doesn't matter and alerts sounds fine 😀

> Contact points contain the configuration for sending notifications. A contact point is a list of integrations each of which sends a notification to a particular email address, service or URL. Contact points can have multiple integrations of the same kind, or a combination of integrations of different kinds. For example, a contact point could contain a Pagerduty integration; an email and Slack integration; or a Pagerduty integration, a Slack integration, and two email integrations. A contact point can also be configured with no integrations in which case no notifications are sent.

Author: But we send an alert to the notification policy, instead of a notification right? The notification policy uses the contact point to turn that alert into a notification.

> However a contact point cannot send notifications until it has been added to a notification policy. A notification policy can only send alerts to one contact point, but a contact point can be added to a number of notification policies at the same time. When an alert matches a notification policy, the alert is sent to the contact point in that notification policy, which then sends a notification to each integration in its configuration.

Author: I'm hesitant to add this suggestion as we haven't used the word entry before to talk about an individual notification policy, and I fear it might confuse the reader. What do you think?

> However a contact point cannot send notifications until it has been added to a notification policy. A notification policy can only send alerts to one contact point, but a contact point can be added to a number of notification policies at the same time. When an alert matches a notification policy, the alert is sent to the contact point in that notification policy, which then sends a notification to each integration in its configuration.

Reviewer: I thought this too, any alternate wording is fine IMO, I just felt the original of:

> A notification policy can only send alerts to one contact point

Is not necessarily true since a policy can have children.

> However a contact point cannot send notifications until it has been added to a notification policy. A notification policy can only send alerts to one contact point, but a contact point can be added to a number of notification policies at the same time. When an alert matches a notification policy, the alert is sent to the contact point in that notification policy, which then sends a notification to each integration in its configuration.

Author: Even when it matches a child policy doesn't only the child policy contact point receive the alert? Or are alerts sent to the contact points of all matching notification policies?

> However a contact point cannot send notifications until it has been added to a notification policy. A notification policy can only send alerts to one contact point, but a contact point can be added to a number of notification policies at the same time. When an alert matches a notification policy, the alert is sent to the contact point in that notification policy, which then sends a notification to each integration in its configuration.

Reviewer: With Continue=true, you can make one policy node appear to dispatch to multiple points, but thats an advanced feature. 🤷 

I think this also suffers from conflating the term here. Does a "notification policy" mean the whole tree? A subtree? A single node? Etc.

  - contact point
  - notification channel
  - create
title: Notifications
weight: 410
---

# Notifications

> Notifications are sent when an alert is firing or has been resolved, and configuring how and where a notification should be sent is done via notification policies. With notification policies it is possible to configure how often a notification should be sent; and whether alerts should all be sent in the same notification, sent in grouped notifications based on a set of labels, or as seperate notifications.

Reviewer 2: ```suggestion
Notifications are sent when an alert is firing or has been resolved. You use notification policies to configure how and where a notification is sent. Configure how often a notification should be sent; whether to send all alerts in the same notification, in grouped notifications based on a set of labels, or as separate notifications.
```

## Notification policies

> Notification policies control when and where notifications are sent. A notification policy can choose to send all alerts together in the same notification, send alerts in grouped notifications based on a set of labels, or send alerts as separate notifications. Each notification policy can have its own timings configuration that control how often notifications should be sent; and can reference one or more mute timings to inhibit notifications at certain times of the day and on certain days of the week.

Reviewer 2: This contains a lot of the text from above - you could include the last part in the paragraph above?

Reviewer 2: Also what about silences?

> Notification policies control when and where notifications are sent. A notification policy can choose to send all alerts together in the same notification, send alerts in grouped notifications based on a set of labels, or send alerts as separate notifications. Each notification policy can have its own timings configuration that control how often notifications should be sent; and can reference one or more mute timings to inhibit notifications at certain times of the day and on certain days of the week.

Reviewer 2: ```suggestion
Notification policies control when and where notifications are sent. A notification policy can choose to send all alerts together in the same notification, send alerts in grouped notifications based on a set of labels, or send alerts as separate notifications. You can configure each notification policy to control how often notifications should be sent as well as adding  one or more mute timings to inhibit notifications at certain times of the day and on certain days of the week.
```

> Notification policies are organized in a tree structure where at the root of the tree there is a notification policy called the root policy. There can be only one root policy, and the root policy cannot be deleted.

Reviewer 2: ```suggestion
Notification policies are organized in a tree structure where at the root of the tree there is a notification policy called the root policy. There can be only one root policy and the root policy cannot be deleted.
```

> Specific routing policies are descendents of the root policy, and can be used to match either all alerts or a subset of alerts based on a set of matching labels. A notification policy matches an alert when its matching labels match the labels in the alert.
> 
> A specific routing policy can have it's own descendent policies, called nested policies, which allow for additional matching of alerts. An example of a specific routing policy could be sending infrastructure alerts to the Ops team; while a descendent policy might send high priority alerts to Pagerduty and low priority alerts as emails.
> 
> All alerts, irrespective of their labels, match the root policy. However, when the root policy receives an alert it looks at each specific routing policy and sends the alert to the first specific routing policy that matches the alert. If the specific routing policy has further descendent policies, then it can attempt to the match the alert against one of its nested policies. If no nested policies match the alert then the specific routing policy is the matching policy. If there are no specific routing policies, or no specific routing policies match the alert, then the root policy is the matching policy.

Reviewer 2: ```suggestion
Specific routing policies are descendents of the root policy and can be used to match either all alerts or a subset of alerts based on a set of matching labels. A notification policy matches an alert when its matching labels match the labels in the alert.

A specific routing policy can have its own descendent policies, called nested policies, which allow for additional matching of alerts. An example of a specific routing policy could be sending infrastructure alerts to the Ops team; while a descendent policy might send high priority alerts to Pagerduty and low priority alerts as emails.

All alerts, irrespective of their labels, match the root policy. However, when the root policy receives an alert it looks at each specific routing policy and sends the alert to the first specific routing policy that matches the alert. If the specific routing policy has further descendent policies, then it can attempt to the match the alert against one of its nested policies. If no nested policies match the alert then the specific routing policy is the matching policy. If there are no specific routing policies, or no specific routing policies match the alert, then the root policy is the matching policy.
```

> 

Reviewer 2: Is descendants the official terminology? is there anything we can change it to to be more user friendly? err...siblings...? not sure.

Author: Siblings is something else, we could call them children?

More information on how to configure notification policies can be found [here]({{< relref "../../manage-notifications/create-notification-policy/" >}}).

## Contact points

> Contact points contain the configuration for sending notifications. A contact point is a list of integrations, each of which sends a notification to a particular email address, service or URL. Contact points can have multiple integrations of the same kind, or a combination of integrations of different kinds. For example, a contact point could contain a Pagerduty integration; an email and Slack integration; or a Pagerduty integration, a Slack integration, and two email integrations. A contact point can also be configured with no integrations in which case no notifications are sent.
> 

Reviewer 2: ```suggestion
Contact points contain the configuration for sending notifications. A contact point is a list of integrations, each of which sends a notification to a particular email address, service or URL. Contact points can have multiple integrations of the same kind, or a combination of integrations of different kinds. For example, a contact point could contain a Pagerduty integration; an email and Slack integration; or a Pagerduty integration, a Slack integration, and two email integrations. You can also configure a contact point with no integrations; in which case no notifications are sent.

```

| [WeCom](#wecom)                                  | `wecom`                   | Supported            | N/A                                                                                                      |
| [Zenduty](https://www.zenduty.com/)              | `webhook`                 | Supported            | N/A                                                                                                      |

## Templating notifications

In Grafana it is possible to customize notifications with templates. For example, templates can be used to change the subject and message of an email, or the title and message of notifications sent to Slack.

> A template is not limited to an individual integration or contact point either, but instead can be used in a number of integrations in the same contact point and even integrations across different contact points. For example, a Grafana user can create a template called `custom_subject_or_title` and use it for both templating subjects in emails and titles of Slack messages without having to create two separate templates.

Author: This is where I think the phrasing used in the UI is not helpful. Somehow we're using the word "contact point type" to mean a configured integration in a contact point, when I think it really should mean the type of integration (i.e. email). You can also have multiple contact point "types" in the same contact point. It's strange when read like that.

However, what I am trying to explain here is that not only can you use the same template in different contact point types (i.e. email and slack), you can also use them in a number of similar "types" in the same contact point, and in different contact points.

I'm considering if we should change it from contact point type to integration because I think the current name is confusing.

> A template is not limited to an individual integration or contact point either, but instead can be used in a number of integrations in the same contact point and even integrations across different contact points. For example, a Grafana user can create a template called `custom_subject_or_title` and use it for both templating subjects in emails and titles of Slack messages without having to create two separate templates.

Reviewer: I struggled on the wording here too, this took me like 3 passes and I agree it's not perfect 😀

I was mainly just trying to shorten it a bit, the original wording felt a little too drawn out, but it's really a minor nit.

| [WeCom](#wecom)                                  | `wecom`                   | Supported            | N/A                                                                                                      |
| [Zenduty](https://www.zenduty.com/)              | `webhook`                 | Supported            | N/A                                                                                                      |

## Templating notifications

> You can to customize notifications with templates. For example, templates can be used to change the subject and message of an email, or the title and message of notifications sent to Slack.
> 
> Templates are not limited to an individual integration or contact point, but instead can be used in a number of integrations in the same contact point and even integrations across different contact points. For example, a Grafana user can create a template called `custom_subject_or_title` and use it for both templating subjects in emails and titles of Slack messages without having to create two separate templates.

Reviewer 2: ```suggestion
You can customize notifications with templates. For example, templates can be used to change the subject and message of an email, or the title and message of notifications sent to Slack.

Templates are not limited to an individual integration or contact point, but instead can be used in a number of integrations in the same contact point and even integrations across different contact points. For example, a Grafana user can create a template called `custom_subject_or_title` and use it for both templating subjects in emails and titles of Slack messages without having to create two separate templates.
```
